### PR TITLE
400 on empty file uid|label|reaper upload

### LINKED ISSUE #993: Disallow empty uploads
Allowing file-less uploads on reaper-focussed endpoints (`/upload/uid`, `/upload/label`, `/upload/reaper`) is unintended and needlessly complicates debugging of reaper tools. I suggest we stop allowing such *empty* uploads.

### PR DESCRIPTION
**Pull request by hkethi002**
Fixes #993 
### Review Checklist

- Tests were added to cover all code changes
- Documentation was added / updated
- Code and tests follow standards in CONTRIBUTING.md

### CODE REVIEW

**Comment by nagem** [comment]
I think we would rather `400` (client messed up) than `500` (we messed up).

File: api/upload.py [hunk 1]
```diff
@@ -95,6 +85,19 @@ def process_upload(request, strategy, container_type=None, id_=None, origin=None
     # Ref docs from placer.TargetedPlacer for details.
     if strategy == Strategy.targeted and len(file_fields) > 1:
```
Comment (hkethi002) [inline api/upload.py]: Should I switch this to a 400 as well?

File: api/upload.py [hunk 2]
```diff
@@ -95,6 +85,19 @@ def process_upload(request, strategy, container_type=None, id_=None, origin=None
...
         raise Exception("Targeted uploads can only send one file")
```
Comment (hkethi002) [inline api/upload.py]: Should I switch this to a 400 as well

**Comment by nagem** [comment]
We'd want to change the existing 500s as well. Inside `placer.py` there is a check state, and we have `requireTarget` and `requireMetadata`, would we also want to write one for `requireFile`? Not 100% sure that's the best way forward but would probably be cleaner.

File: api/placer.py [hunk 3]
```diff
@@ -75,7 +77,15 @@ def requireMetadata(self):
         Helper function that throws unless metadata was provided.
         """
         if self.metadata == None:
-            raise Exception('Metadata required')
+            raise files.FileFormException('Metadata required')
+
+    def requireFiles(self):
+        """
+        Helper function that throws unless metadata was provided.
```
Comment (nagem) [inline api/placer.py]: I think you meant this to say "unless no files were uploaded" or similar

File: api/placer.py [hunk 4]
```diff
@@ -114,6 +124,16 @@ class TargetedPlacer(Placer):
     An exception is thrown in upload.process_upload() if you try. This could be fixed by making a better schema.
     """
 
+    # TODO: Change schemas to enabled targeted uploads of more than one file.
+    # Ref docs from placer.TargetedPlacer for details.
+    def requireFiles(self):
```
Comment (nagem) [inline api/placer.py]: Maybe I'm missing it but it doesn't look like this is getting used in the `check()` phase for `TargetedPlacer`

**Comment by kofalt** [comment]
Our of band, Harsha: we can't access `file_fields` that way because it breaks the ability of the Placer interface to incrementally process a metadata stream. Megan and I worked on some alternatives, check with us.

**Comment by nagem** [comment]
@hkethi002 and I went through moving the check back into the `process_upload` path.

File: api/upload.py [hunk 5]
```diff
@@ -95,7 +93,11 @@ def process_upload(request, strategy, container_type=None, id_=None, origin=None
     # TODO: Change schemas to enabled targeted uploads of more than one file.
     # Ref docs from placer.TargetedPlacer for details.
     if strategy == Strategy.targeted and len(file_fields) > 1:
-        raise Exception("Targeted uploads can only send one file")
+        raise FileFormException("Targeted uploads can only send one file")
+    elif strategy in [Strategy.reaper, Strategy.uidupload, Strategy.labelupload, Strategy.uidmatch] and len(file_fields) < 1:
```
Comment (kofalt) [inline api/upload.py]: If you want to keep the incremental processing window open for these upload strategies, then we cannot access `file_fields` here. The fix is to instead have the check during each strategy's `finalize` calls.

Megan and I discussed a possible fix for the targeted case as well, but probably not as part of this PR.

File: api/upload.py [hunk 6]
```diff
@@ -84,8 +84,6 @@ def process_upload(request, strategy, container_type=None, id_=None, origin=None
 
     placer_class = strategy.value
     placer = placer_class(container_type, container, id_, metadata, timestamp, origin, context)
-    placer.check()
```
Comment (kofalt) [inline api/upload.py]: This line shouldn't have moved. Otherwise, this PR seems fine.

**Pull request merged**